pallets/stream-payment/src/lib.rs

```diff
@@ -0,0 +1,978 @@
+                config,
+                deposit: initial_deposit,
+                last_time_updated: now,
+                request_nonce: 0,
+                pending_request: None,
+            };
+
+            // Insert stream in storage.
+            Streams::<T>::insert(stream_id, stream);
+            LookupStreamsWithSource::<T>::insert(origin, stream_id, ());
+            LookupStreamsWithTarget::<T>::insert(target, stream_id, ());
+
+            // Emit event.
+            Pallet::<T>::deposit_event(Event::<T>::StreamOpened { stream_id });
+
+            Ok(stream_id)
+        }
+
+        /// Behavior:
+        /// A stream payment consist of a locked deposit, a rate per unit of time and the
+        /// last time the stream was updated. When updating the stream, **at most**
+        /// `elapsed_time * rate` is unlocked from the source account and transfered to the target
+        /// account. If this amount is greater than the left deposit, the stream is considered
+        /// drained **but not closed**. The source can come back later and refill the stream,
+        /// however there will be no retroactive payment for the time spent as drained.
+        /// If the stream payment is used to rent a service, the target should pause the service
+        /// while the stream is drained, and resume it once it is refilled.
+        fn perform_stream_payment(
+            stream_id: T::StreamId,
+            stream: &mut StreamOf<T>,
+        ) -> Result<T::Balance, DispatchErrorWithPostInfo> {
+            let mut now = T::TimeProvider::now(&stream.config.time_unit)
+                .ok_or(Error::<T>::CantFetchCurrentTime)?;
+
+            // We want to update `stream.last_time_updated` to `now` as soon
+            // as possible to avoid forgetting to do it. We copy the old value
+            // for payment computation.
+            let last_time_updated = stream.last_time_updated;
+            stream.last_time_updated = now;
+
+            // Take into account mandatory change request deadline. Note that
+            // while it'll perform payment up to deadline,
+            // `stream.last_time_updated` is still the "real now" to avoid
+            // retroactive payment in case the deadline changes.
+            if let Some(ChangeRequest {
+                kind: ChangeKind::Mandatory { deadline },
+                ..
+            }) = &stream.pending_request
+            {
+                now = min(now, *deadline);
+            }
+
+            // Dont perform payment
+            if now == last_time_updated {
+                return Ok(0u32.into());
+            }
+
+            // If deposit is zero the stream is fully drained and there is nothing to transfer.
+            if stream.deposit.is_zero() {
+                return Ok(0u32.into());
+            }
+
+            let delta = now
+                .checked_sub(&last_time_updated)
+                .ok_or(Error::<T>::TimeMustBeIncreasing)?;
```

Comment:
Making a mandatary change with `deadline = 0` can result in this error. Can be fixed by enforcing the deadline to be `>= last_time_updated`, either in this function or when calling `request_change`.